core/templates/cowdata.h

```diff
@@ -293,143 +304,137 @@ void CowData<T>::_unref() {
+			const Error error = _realloc(alloc_size);
+			if (error) {
+				// Out of memory; the current array is still valid though.
+				return error;
 			}
 		}
+	} else {
+		// Resize by forking.
+
+		// Create a temporary CowData to hold ownership over our _ptr.
+		// It will be used to copy elements from the old buffer over to our new buffer.
+		// At the end of the block, it will be automatically destructed by going out of scope.
+		const CowData prev_data;
+		prev_data._ptr = _ptr;
+		_ptr = nullptr;
 
-		_unref();
-		_ptr = _data_ptr;
+		const Error error = _alloc(alloc_size);
+		if (error) {
+			return error;
+		}
 
-		rc = 1;
+		// Copy over elements.
+		const USize copied_element_count = MIN(prev_size, p_size);
+		if (copied_element_count > 0) {
+			if constexpr (std::is_trivially_copyable_v<T>) {
+				memcpy((uint8_t *)_ptr, (uint8_t *)prev_data._ptr, copied_element_count * sizeof(T));
+			} else {
+				for (USize i = 0; i < copied_element_count; i++) {
+					memnew_placement(&_ptr[i], T(prev_data._ptr[i]));
+				}
+			}
+		}
 	}
-	return rc;
+
+	// Set our new size.
+	*_get_size() = p_size;
+
+	return OK;
 }
 
 template <typename T>
 template <bool p_ensure_zero>
 Error CowData<T>::resize(Size p_size) {
 	ERR_FAIL_COND_V(p_size < 0, ERR_INVALID_PARAMETER);
 
-	Size current_size = size();
-
-	if (p_size == current_size) {
+	const Size prev_size = size();
+	if (p_size == prev_size) {
 		return OK;
 	}
 
-	if (p_size == 0) {
-		// Wants to clean up.
-		_unref(); // Resets _ptr to nullptr.
-		return OK;
+	const Error error = _fork_allocate(p_size);
+	if (error) {
+		return error;
 	}
 
-	// possibly changing size, copy on write
-	_copy_on_write();
-
-	USize current_alloc_size = _get_alloc_size(current_size);
-	USize alloc_size;
-	ERR_FAIL_COND_V(!_get_alloc_size_checked(p_size, &alloc_size), ERR_OUT_OF_MEMORY);
-
-	if (p_size > current_size) {
-		if (alloc_size != current_alloc_size) {
-			if (current_size == 0) {
-				// alloc from scratch
-				uint8_t *mem_new = (uint8_t *)Memory::alloc_static(alloc_size + DATA_OFFSET, false);
-				ERR_FAIL_NULL_V(mem_new, ERR_OUT_OF_MEMORY);
-
-				SafeNumeric<USize> *_refc_ptr = _get_refcount_ptr(mem_new);
-				USize *_size_ptr = _get_size_ptr(mem_new);
-				T *_data_ptr = _get_data_ptr(mem_new);
-
-				new (_refc_ptr) SafeNumeric<USize>(1); //refcount
-				*(_size_ptr) = 0; //size, currently none
-
-				_ptr = _data_ptr;
-
-			} else {
-				const Error error = _realloc(alloc_size);
-				if (error) {
-					return error;
-				}
-			}
-		}
-
-		// construct the newly created elements
-
+	if (p_size > prev_size) {
 		if constexpr (!std::is_trivially_constructible_v<T>) {
-			for (Size i = *_get_size(); i < p_size; i++) {
+			for (Size i = prev_size; i < p_size; i++) {
 				memnew_placement(&_ptr[i], T);
 			}
 		} else if (p_ensure_zero) {
-			memset((void *)(_ptr + current_size), 0, (p_size - current_size) * sizeof(T));
+			memset((void *)&_ptr[prev_size], 0, (p_size - prev_size) * sizeof(T));
```

Comment:
would have stick with `(void *)(_ptr + prev_size)` but that's style, so both will do

Reply:
I don't feel strongly about this, IIRC i adjusted this to be on-par with other lines using the `[]` syntax.